File: 프로그래머스/3/81303. 표 편집/표 편집.py
# n : 처음 표의 행 개수
# k : 처음에 선택된 행의 위치
# cmd : 수행한 명령어들이 담긴 문자열 배열
# 모든 명령어를 수행한 후 표의 상태와 처음 주어진 표의 상태를 비교하여 삭제되지 않은 행 O, X로 표시하여 리턴
def solution(n, k, cmd):
    answer = ['O'] * n
    prev = {i: i-1 for i in range(n)}
    next = {i: i+1 for i in range(n)}
    cancels = []
    
    for c in cmd:
        if c[0] == "U":
            cnt = int(c.split()[1])
            for _ in range(cnt):
                k = prev[k]
        elif c[0] == 'D':
            cnt = int(c.split()[1])
            for _ in range(cnt):
                k = next[k]
        elif c[0] == 'C':
            answer[k] = 'X'
            cancels.append((k, prev[k], next[k])) # 현재 행과 이전/다음 정보를 저장
            
            # 이전과 다음을 서로 연결하여 삭제된 행을 건너뜀
            if prev[k] != -1:
                next[prev[k]] = next[k]
            if next[k] != n:
                prev[next[k]] = prev[k]
                
            # 마지막 행이면 커서를 위로, 그렇지 않으면 아래로 이동
            k = next[k] if next[k] != n else prev[k]
        elif c[0] == 'Z':
            idx, p, n_ = cancels.pop()
            answer[idx] = 'O'
            
            # 복구된 행을 다시 이전과 다음에 연결
            # p가 삭제된 상태일 경우, p의 값이 존재하는 첫번재 행을 찾음
            while p != -1 and answer[p] == 'X':
                p = prev[p]
            if p != -1:
                next[p] = idx
            
            # n_ 이 삭제된 상태일 경우, n_의 값이 존재하는 첫번째 행을 찾음
            while n_ != n and answer[n_] == 'X':
                n_ = next[n_]
            if n_ != n:
                prev[n_] = idx
                
            prev[idx] = p
            next[idx] = n_
                
    return ''.join(answer)

File: 프로그래머스/3/81303. 표 편집/표 편집.java
import java.util.*;

class Solution {
    public String solution(int n, int k, String[] cmd) {
        char[] answer = new char[n];
        Arrays.fill(answer, 'O');
        Node cur = null;
        Node prev = null; 
        ArrayDeque<Node> trash = new ArrayDeque<>();
        
        for (int i = 0;i<n;i++){
            Node node = new Node(prev, null, i);
            if(prev != null) {
                prev.setNext(node);   
            }
            if (i == k) {
                cur = node;
            }
            prev = node;
        }
        
        for (String c : cmd) {
            if (c.split(" ")[0].equals("U")) {
                int x = Integer.parseInt(c.split(" ")[1]);
                
                for (int i=0;i<x;i++){
                    cur = cur.getPrev();
                }
            } else if(c.split(" ")[0].equals("D")) {
                int x = Integer.parseInt(c.split(" ")[1]);
                
                for (int i=0;i<x;i++){
                    cur = cur.getNext();
                }
            } else if(c.equals("C")) {
                trash.push(cur);
                answer[cur.getValue()] = 'X';
                if(cur.getNext() != null) {
                    if(cur.getPrev()!= null) {
                        cur.getPrev().setNext(cur.getNext());    
                    }
                    cur.getNext().setPrev(cur.getPrev());
                    cur = cur.getNext();
                } else {
                    cur.getPrev().setNext(null);
                    cur = cur.getPrev();
                }
            } else { // Z
                Node re = trash.pop();
                answer[re.getValue()] = 'O';
                if(re.getPrev()!=null) {
                    re.getPrev().setNext(re);    
                }
                if(re.getNext()!=null){
                    re.getNext().setPrev(re);    
                }
            }
        }
        
        return new String(answer);
    }
    
    public class Node{
        Node prev;
        int value;
        Node next;
        
        Node(Node prev, Node next, int value) {
            this.prev = prev;
            this.next = next;
            this.value = value;
        }
        
        public void setPrev(Node prev) {
            this.prev = prev;
        }
        
        public void setNext(Node next) {
            this.next =next;
        }
        
        public Node getPrev() {
            return prev;
        }
        
        public Node getNext() {
            return next;
        }
        
        public int getValue() {
            return value;
        }
    }
}